### PR TITLE
chore(tracing): add tracing for frontend and db session

### PR DESCRIPTION
**Pull request by jalevin**
**What is this feature?**
This PR adds instrumentation for loading frontend SPA along with select methods in the dashboard service, and cleans up span handling in sqlstore.

### CODE REVIEW

File: pkg/api/dashboard.go [hunk 1]
```diff
@@ -82,7 +82,7 @@ func dashboardGuardianResponse(err error) response.Response {
 // 404: notFoundError
 // 500: internalServerError
 func (hs *HTTPServer) GetDashboard(c *contextmodel.ReqContext) response.Response {
-	ctx, span := hs.tracer.Start(c.Req.Context(), "httpserver.GetDashboard")
+	ctx, span := hs.tracer.Start(c.Req.Context(), "api.GetDashboard")
```
Comment (hairyhenderson) [inline pkg/api/dashboard.go:85]: 🤔 I'm not sure this rename is necessary - this is the `api` package, but the `HTTPServer` type... the naming idiom I've usually used is to use `<type>.<method>`... though I could go either way...

File: pkg/api/utils.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ import (
 	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
 	"github.com/grafana/grafana/pkg/services/login"
 	"github.com/grafana/grafana/pkg/services/user"
+	trace "go.opentelemetry.io/otel/trace"
```
Comment (hairyhenderson) [inline pkg/api/utils.go:14]:
```suggestion
	"go.opentelemetry.io/otel/trace"
```

File: pkg/services/search/service.go [hunk 3]
```diff
@@ -14,7 +15,7 @@ import (
 	"github.com/grafana/grafana/pkg/setting"
 )
 
-func ProvideService(cfg *setting.Cfg, sqlstore db.DB, starService star.Service, dashboardService dashboards.DashboardService) *SearchService {
+func ProvideService(cfg *setting.Cfg, sqlstore db.DB, starService star.Service, dashboardService dashboards.DashboardService, tracer tracing.Tracer) *SearchService {
```
Comment (hairyhenderson) [inline pkg/services/search/service.go:18]: Can we use a package-level tracer instead? IMO it's a much simpler pattern to use, and we get a useful component name out of it:

```suggestion
var tracer = otel.Tracer("github.com/grafana/grafana/pkg/services/search")

func ProvideService(cfg *setting.Cfg, sqlstore db.DB, starService star.Service, dashboardService dashboards.DashboardService) *SearchService {
```

File: pkg/api/dashboard.go (hunk 1 above)
Comment (jalevin) [reply pkg/api/dashboard.go:85]: After discussion we decided to keep <package>.<method> because of the way we use interfaces across the codebase

**Review by hairyhenderson — APPROVED** [review]
LGTM, thanks!

**Comment by mildwonkey** [comment]
I'm a little unsure about the sql spans, do you have an example of what the new output looks like? Something from a local instance or using deploy-to-hg?

File: pkg/api/utils.go [hunk 4]
```diff
@@ -63,3 +64,9 @@ func ValidateAndNormalizeEmail(email string) (string, error) {
 
 	return e.Address, nil
 }
+
+func (hs *HTTPServer) injectSpan(c *contextmodel.ReqContext, name string) (*contextmodel.ReqContext, trace.Span) {
+	ctx, span := hs.tracer.Start(c.Req.Context(), name)
+	c.Req = c.Req.WithContext(ctx)
+	return c, span
+}
```
Comment (diegommm) [inline pkg/api/utils.go:72]: Would it make more sense to add a middleware instead or do we want to keep it only for specific endpoints? Do we also want to add other information we can gather to the span as well?

Comment (jalevin) [reply pkg/api/utils.go:72]: Right now we want to keep it limited to specific endpoints and figure out our usecase. We can bolt onto this as we go and think about what a more generic middleware might look like. Having a span on api failures would be nice.

**Comment by jalevin** [comment]
> I'm a little unsure about the sql spans, do you have an example of what the new output looks like? Something from a local instance or using deploy-to-hg?

Here's what the new trace lifecycle looks like for middleware to frontend.
image

I'm trying to figure out a contrived way to create a db error that isn't done with load testing. It may be easier to ship and circle back to see if we like the approach or not.

**Pull request merged**